### PR TITLE
Performance regression in stat_ops.FrameMultiIndexOps.time_op

### LINKED ISSUE #35050: Performance regression in stat_ops.FrameMultiIndexOps.time_op
https://pandas.pydata.org/speed/pandas/index.html#stat_ops.FrameMultiIndexOps.time_op?p-level=1&p-op=%27std%27&commits=c9144ca54dcc924995acae3d9dcb890a5802d7c0

```python
import pandas as pd
import numpy as np

levels = [np.arange(10), np.arange(100), np.arange(100)]
codes = [
    np.arange(10).repeat(10000),
    np.tile(np.arange(100).repeat(100), 10),
    np.tile(np.tile(np.arange(100), 100), 10),
]
index = pd.MultiIndex(levels=levels, codes=codes)
df = pd.DataFrame(np.random.randn(len(index), 4), index=index)
%timeit df.groupby(level=1).std()
```

Points to https://github.com/pandas-dev/pandas/pull/34372 (cc @rhshadrach), but there was an earlier slowdown.

### PR DESCRIPTION
**Pull request by simonjayhawkins**
- [ ] closes #35050
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

```
import pandas as pd
import numpy as np

levels = [np.arange(10), np.arange(100), np.arange(100)]
codes = [
    np.arange(10).repeat(10000),
    np.tile(np.arange(100).repeat(100), 10),
    np.tile(np.tile(np.arange(100), 100), 10),
]
index = pd.MultiIndex(levels=levels, codes=codes)
df = pd.DataFrame(np.random.randn(len(index), 4), index=index)
%timeit df.std(level=1)
# 9.09 ms ± 59.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each) -> master
# 7.39 ms ± 71.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each) -> 0.25.3
# 7.21 ms ± 113 µs per loop (mean ± std. dev. of 7 runs, 100 loops each) -> PR
```

### CODE REVIEW

File: pandas/core/groupby/groupby.py [hunk 1]
```diff
@@ -1434,16 +1434,18 @@ def std(self, ddof: int = 1):
...
-            "group_var_float64",
-            aggregate=True,
-            needs_counts=True,
-            needs_values=True,
-            needs_2d=True,
-            cython_dtype=np.dtype(np.float64),
-            post_processing=lambda vals, inference: np.sqrt(vals),
-            ddof=ddof,
-        )
+        result = self.var(ddof=ddof)
```
Comment (jreback) [inline pandas/core/groupby/groupby.py:1437]: hmm you are reverting.

instead I would like to see if the generic function can be patched (e.g. something is fundamentally different).

cc @rhshadrach

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/groupby.py (hunk 1 above)
Comment (rhshadrach) [reply pandas/core/groupby/groupby.py:1437]: I don't see a way. `_get_cythonized_result` operates column-by-column, creating overhead. On data this size, we spend 25% in `grouper.group_info` and 40% in `_wrap_aggregated_output`. Within each of these operations, I don't see any easy wins. Only 27% of the time is spent doing the actual computation. 

The only way I see to improve performance is to change `_get_cythonized_result` to a 2d operation, operating on all columns at once. In the PR that caused this regression, that's the first thing I tried and found that it got too hairy. In the process, I realized that perhaps a better solution would be to incorporate the features of `_get_cythonized_result` (namely, pre- and post-processing) into `_cython_agg_general`.

Comment (simonjayhawkins) [reply pandas/core/groupby/groupby.py:1437]: > hmm you are reverting.

yep. should have mentioned this in the OP.



> instead I would like to see if the generic function can be patched (e.g. something is fundamentally different).

OK. will close this for now. can reopen if a better solution is not forthcoming before the release.

Comment (simonjayhawkins) [reply pandas/core/groupby/groupby.py:1437]: > OK. will close this for now. can reopen if a better solution is not forthcoming before the release.

or maybe better would be to backport this after 1.1rc if not fixed before.

**Comment by simonjayhawkins** [comment]
@jreback if we're not branching, should I reopen this and merge to master? see https://github.com/pandas-dev/pandas/issues/34730#issuecomment-656182086

**Comment by jreback** [comment]
no reverting is not useful here

**Comment by rhshadrach** [comment]
If we're not reverting, the only three options I can see are:

1. Let the regression stand.
2. Make _get_cythonized_result a 2d computation rather than operating on each column (which I first tried, and backed out once I realized how difficult this would be). I also expect that doing so would also involve a performance regression for single-column frames that go through this path.
3. Add pre/post-processing to _cython_agg_general.

I don't know how viable/difficult 3 would be, and I'd be happy to work on it, but I don't think we want it in 1.1.0 at this point even if it could get done.

**Pull request closed without merging**